runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -1190,6 +1190,18 @@ J9::Z::TreeEvaluator::inlineVectorizedStringIndexOf(TR::Node* node, TR::CodeGene
       cursor = generateVRScInstruction(cg, TR::InstOpCode::VLGV, node, matchIndexReg, searchResultVReg, generateS390MemoryReference(7, cg), 0);
       iComment("check 7th index of search result vec for byte index");
       generateRRInstruction(cg, TR::InstOpCode::getAddRegOpCode(), node, matchIndexReg, stringIndexReg);
+
+      // An edge case failure can occur when the string we would like to search in is smaller than 16 characters and -XX:+CompactStrings is enabled.
+      // The scenario can be illustrated by the example below:
+      //    If V2 = "teststring", then vector representation of V2 = {116, 101, 115, 116, 83, 116, 114, 105, 110, 103, 0, 0, 0, 0, 0, 0}.
+      //    If V3 = "\0", then vector representation of V3 = {0}.
+      //    Therefore, V4 = {0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0}. (i.e. substring length is 1).
+
+      // In the above example, V1 would have the result 10 (i.e. pointing to the first occurrence of 0 in V2). This is obviously incorrect as an index
+      // value of 10 is out of range in the string we are searching.
+      // Therefore, we conditionally jump to labelPatternNotFound using the instruction below if matchIndexReg > maxIndexReg.
+      cursor = generateRIEInstruction(cg, TR::InstOpCode::getCmpRegAndBranchRelOpCode(), node, matchIndexReg, maxIndexReg, labelPatternNotFound, TR::InstOpCode::COND_BH);
+      iComment("Jump if pattern match found beyond end of string (i.e. we matched garbage data in vector residue).")
```

Comment:
There's a semicolon missing here

Reply:
Sorry to be picky. But I think the term "garbage data" might be a bit confusing because VLL writes 0s to all unused vector register slots. Maybe we could say 0s from unused vector register slots. This evaluator is pretty complex as it is so every little bit helps 🙂

Reply:
Thanks for catching that! These should be fixed in the latest commit 8ff182d